feat: Add auto-instrumentation support for AWS resources in AWS SDK v1

lukeina2z opened the pull request:
This PR adds auto-instrumentation support for the following AWS resources:

Lambda
Secrets Manager
SNS
Step Functions

Tests Run:
./gradlew spotlessCheck
./gradlew clean assemble
./gradlew instrumentation:test
./gradlew :smoke-tests:test

No regression issues found. All newly added tests pass.

Backward Compatibility:
There is no risk of breaking existing functionality. This change only adds instrumentation for additional AWS resources without modifying the existing behavior of the auto-instrumentation library.

laurit commented on instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/RequestAccess.java:
```diff
@@ -90,6 +150,12 @@ private RequestAccess(Class<?> clz) {
+    getStateMachineArn = findAccessorOrNull(clz, "getStateMachineArn");
+    getStepFunctionsActivityArn = findAccessorOrNull(clz, "getActivityArn");
+    getSnsTopicArn = findAccessorOrNull(clz, "getTopicArn");
+    getSecretArn = findAccessorOrNull(clz, "getARN");
```
`getARN` is also present in https://github.com/aws/aws-sdk-java/blob/76838885247a94fc6b0acc49c46f718473d38e63/aws-java-sdk-wafv2/src/main/java/com/amazonaws/services/wafv2/model/GetRuleGroupRequest.java#L377 and in some other classes. Perhaps you need to find a way to only read this from some specific types? Or perhaps there should be a generic arn attribute?

lukeina2z replied:
Good feedback. I will add conditional check with the class name.

laurit commented on instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/RequestAccess.java:
```diff
@@ -90,6 +150,12 @@ private RequestAccess(Class<?> clz) {
     getTargetArn = findAccessorOrNull(clz, "getTargetArn");
+    getStateMachineArn = findAccessorOrNull(clz, "getStateMachineArn");
+    getStepFunctionsActivityArn = findAccessorOrNull(clz, "getActivityArn");
+    getSnsTopicArn = findAccessorOrNull(clz, "getTopicArn");
```
there is already `getTopicArn` that is used to fill `messaging.destination.name` attribute for sns messages

laurit commented on instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/RequestAccess.java:
```diff
@@ -90,6 +150,12 @@ private RequestAccess(Class<?> clz) {
+    getStepFunctionsActivityArn = findAccessorOrNull(clz, "getActivityArn");
+    getSnsTopicArn = findAccessorOrNull(clz, "getTopicArn");
+    getSecretArn = findAccessorOrNull(clz, "getARN");
+    getLambdaName = findAccessorOrNull(clz, "getFunctionName");
```
Also used in https://github.com/aws/aws-sdk-java/blob/76838885247a94fc6b0acc49c46f718473d38e63/aws-java-sdk-glue/src/main/java/com/amazonaws/services/glue/model/GetUserDefinedFunctionRequest.java#L155 that doesn't seem to be related to lambdas.

lukeina2z replied:
Will address this.

laurit commented on instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/RequestAccess.java:
```diff
@@ -90,6 +150,12 @@ private RequestAccess(Class<?> clz) {
+    getSnsTopicArn = findAccessorOrNull(clz, "getTopicArn");
+    getSecretArn = findAccessorOrNull(clz, "getARN");
+    getLambdaName = findAccessorOrNull(clz, "getFunctionName");
+    getLambdaResourceId = findAccessorOrNull(clz, "getUUID");
```
I couldn't find it being used outside of lambda sdk, but given its generic name it could benefit from an extra check to ensure that it is not added to some random request

lukeina2z commented on instrumentation/aws-sdk/aws-sdk-1.11/library/build.gradle.kts:
```diff
@@ -17,6 +17,9 @@ dependencies {
   testLibrary("com.amazonaws:aws-java-sdk-kinesis:1.11.106")
   testLibrary("com.amazonaws:aws-java-sdk-dynamodb:1.11.106")
   testLibrary("com.amazonaws:aws-java-sdk-sns:1.11.106")
+  testLibrary("com.amazonaws:aws-java-sdk-secretsmanager:1.11.309")
+  testLibrary("com.amazonaws:aws-java-sdk-stepfunctions:1.11.230")
+  testLibrary("com.amazonaws:aws-java-sdk-lambda:1.11.678")
```
Thank you for the feedback. I will keep the version number consistent with the existing dependency.

lukeina2z commented:
@laurit I addressed all the review comments. Would you please review the new iteration? Sorry I made a rebase from main branch, which makes the diff between 1st and 2nd iterations a bit out of sync.

lukeina2z commented:
I am going to split this into smaller chunks to make code review easier.
